### PR TITLE
[FSDP] Add state_dict() save/reload in parity test

### PR DESCRIPTION
**Pull request by rohan-varma**
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #73532
* #73453
* __->__ #73366

Adds state_dict() save/reload in parity with DDP test to ensure
checkpointing doesn't cause issue with accuracy/model params.

Differential Revision: [D34434358](https://our.internmc.facebook.com/intern/diff/D34434358/)

### CODE REVIEW

**Review by fegin — APPROVED** [review]
Thanks for the test.
